test/e2e_node/docker_test.go

```diff
@@ -70,4 +78,170 @@ var _ = framework.KubeDescribe("Docker features [Feature:Docker]", func() {
+			// the container is running by checking if the timestamp increases.
+			cmd := "" +
+				"while true; do " +
+				"    date +%s > /test-dir/" + timestampFilename + "; " +
+				"    sleep 1; " +
+				"done"
+			By("Create the test pod.")
+			f.PodClient().CreateSync(&v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{Name: podName},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:    containerName,
+						Image:   "gcr.io/google_containers/busybox:1.24",
+						Command: []string{"/bin/sh"},
+						Args:    []string{"-c", cmd},
+						VolumeMounts: []v1.VolumeMount{
+							{
+								Name:      volumeName,
+								MountPath: "/test-dir",
+							},
+						},
+					}},
+					Volumes: []v1.Volume{{
+						Name: volumeName,
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{Path: tempDir},
+						},
+					}},
+				},
+			})
+
+			startTime1, err := getContainerStartTime(f, podName, containerName)
+			framework.ExpectNoError(err)
+
+			By("Stop Docker daemon.")
+			framework.ExpectNoError(stopDockerDaemon())
+			defer func() {
+				By("Restart Docker daemon.")
+				framework.ExpectNoError(startDockerDaemon())
+			}()
+
+			By("Ensure that the test container is running when Docker daemon is down.")
+			isRunning, err := isContainerRunning(filepath.Join(tempDir, timestampFilename))
+			framework.ExpectNoError(err)
+			if !isRunning {
+				framework.Failf("The container should be running but it's not.")
+			}
+
+			By("Ensure that the test pod is still running when Docker daemon is down.")
+			framework.ExpectNoError(f.WaitForPodRunning(podName))
+
+			By("Start Docker daemon.")
+			framework.ExpectNoError(startDockerDaemon())
+
+			By("Ensure that the test container is running after Docker daemon is restarted.")
+			isRunning, err = isContainerRunning(filepath.Join(tempDir, timestampFilename))
+			framework.ExpectNoError(err)
+			if !isRunning {
+				framework.Failf("The container should be running but it's not.")
+			}
+
+			By("Ensure that the test container has not been restarted after Docker daemon is restarted.")
+			startTime2, err := getContainerStartTime(f, podName, containerName)
+			framework.ExpectNoError(err)
+			if startTime1 != startTime2 {
+				framework.Failf("The container should have not been restarted.")
+			}
+		})
+	})
 })
+
+// isContainerRunning returns true if the container is running (by checking
+// whether the timestamp is being updated), and false otherwise. Returns an
+// error if the timestamp cannot be read.
+func isContainerRunning(filename string) (bool, error) {
+	const (
+		// The sample interval (3s), which must be greater than the interval at
+		// which the container writes the timestamp (every second).
+		sampleInterval = 3 * time.Second
+		retryInterval  = 3 * time.Second
+		retryTimeout   = 30 * time.Second
+	)
+	for start := time.Now(); time.Since(start) < retryTimeout; time.Sleep(retryInterval) {
```

Comment:
What are we retrying here?

We should make sure the container `Consistently` running for a period of time.